### DISCUSSION TITLE
jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type

### DISCUSSION THREAD

**Discussion by nicolasduminil** [quarkusio/quarkus#33343] [Q&A] [answered]
Hello,

I have a basic CDI example with Quarkus 3.0.3.Final that I'm not able to get working. The reproducer is [here](https://github.com/nicolasduminil/quarkus-cdi-test).

Basically, I have a maven multi-module project in which there is an API as below:

    public interface QuarkusCDIFacade
    {
      String sayHello (String toWhom);
    }

a provider as follows:

    @ApplicationScoped
    public class DefaultQuarkusCDIFacade implements QuarkusCDIFacade
    {
      @Override
      public String sayHello(String toWhom)
      {
        return "Hello " + toWhom;
      }
    }

and a consumer like this:

    @ApplicationScoped
    public class QuarkusCDIFacadeConsumer
    {
      @Inject
      QuarkusCDIFacade uarkusCDIFacade;
    }

When I try to build, I get:

    [ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.0.3.Final:build (default) on project cdi-consumer: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
    [ERROR]         [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type fr.simplex_software.quarkus.cdi.api.QuarkusCDIFacade and qualifiers [@Default]
    [ERROR]         - java member: fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer#uarkusCDIFacade
    [ERROR]         - declared on CLASS bean [types=[fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer, java.lang.Object], qualifiers=[@Default, @Any], target=fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer]

I'm using CDI since years, not with Quarkus but with Jakarta EE, and I've never had such an issue. Do I miss something here ?

Many thanks in advance.

Kind regards,

Nicolas

**Comment by maxandersen** [comment]
Look for jandex index in https://quarkus.io/guides/cdi-reference

That should explain and provide info to set it up.

**Reply by nicolasduminil** [reply]
If you have looked in the reproducer project, you would have seen that the jandex maven plugin is configured as expected.

**Reply by mschorsch** [reply]
`cdi-api` should also be indexed by the maven-jandex-plugin.

**Reply by nicolasduminil** [reply]
Have updated cdi-api pom.xml such that to add the maven-jandex-plugin. It doesn't change anything, as shown below:

    [ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.0.3.Final:build (default) on project cdi-consumer:    Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
    [ERROR]         [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type fr.simplex_software.quarkus.cdi.api.QuarkusCDIFacade and qualifiers [@Default]
     [ERROR]         - java member: fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer#uarkusCDIFacade
    [ERROR]         - declared on CLASS bean [types=[fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer, java.lang.Object], qualifiers=[@Default, @Any], target=fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer]
    [ERROR]         at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1395)
    [ERROR]         at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:305)
    [ERROR]         at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:156)
    [ERROR]         at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:526)
    [ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    [ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    [ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
    [ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
    [ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
    [ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
    [ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
    [ERROR]         at java.base/java.lang.Thread.run(Thread.java:834)
    [ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
    [ERROR] Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type fr.simplex_software.quarkus.cdi.api.QuarkusCDIFacade and qualifiers [@Default]
    [ERROR]         - java member: fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer#uarkusCDIFacade
    [ERROR]         - declared on CLASS bean [types=[fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer, java.lang.Object], qualifiers=[@Default, @Any], target=fr.simplex_software.quarkus.cdi.consumer.QuarkusCDIFacadeConsumer]
    [ERROR]         at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:482)
    [ERROR]         at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:564)
    [ERROR]         at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:293)

I don't understand how is that possible to spend one week trying to get working the simplest possible test case !

**Reply by gsmet** [reply, marked as answer]
> I don't understand how is that possible to spend one week trying to get working the simplest possible test case !

It certainly doesn't help that there is multiple typos in your reproducer. Please make it compile and then we could have a look at what you did.

I fixed:
```diff
diff --git a/cdi-provider/pom.xml b/cdi-provider/pom.xml
index c687105..d680556 100644
--- a/cdi-provider/pom.xml
+++ b/cdi-provider/pom.xml
@@ -16,7 +16,7 @@
       <artifactId>quarkus-arc</artifactId>
     </dependency>
     <dependency>
-      <groupId>fr.simplex_spftware.quarkus.cdi</groupId>
+      <groupId>fr.simplex-software.quarkus.cdi</groupId>
       <artifactId>cdi-api</artifactId>
       <version>1.0-SNAPSHOT</version>
     </dependency>
@@ -34,4 +34,4 @@
       </plugin>
     </plugins>
   </build>
-</project>
\ No newline at end of file
+</project>
```

but there are other errors.

**Reply by nicolasduminil** [reply]
Thank you for pointing me to these typos. They were the cause of my issue. But as you noticed, correcting them still didn't compile and raised the following exception:

    [ERROR] COMPILATION ERROR : 
    [INFO] -------------------------------------------------------------
    [ERROR] /home/nicolas/QuarkusCDI/cdi-provider/src/main/java/fr/simplex_software/quarkus/cdi/provider/DefaultQuarkusCDIFacade.java:[6,84] package fr.simplex_software.quarkus.cdi.api does not exist
    [ERROR] /home/nicolas/QuarkusCDI/cdi-provider/src/main/java/fr/simplex_software/quarkus/cdi/provider/DefaultQuarkusCDIFacade.java:[8,3] method does not override or implement a method from a supertype

These compilation errors didn't make any sense as of course that the mentioned package existed. After wasting some time on that, I discovered something of unbelievable. Commenting out the following line in the cdi-api module:

      <name>QuarkusCDI :: The API module</name>

the project compiles correctly. Uncommenting it back, the compilation fails again.

So, in order to resume: my issue has been solved by the fact of adding the maven-jandex-plugin to the cdi-api module. Initially, this plugin was configured for the cdi-provider and cdi-consumer modules. I thought that this was enough but obviously it wasn't and I had to do it for cdi-api as well. This is what fixed the issue.

But by doing that, I came to a code that didn't compile, for incomprehensible reasons. This was a 2ns issue. Andn as unbeleivable as it might seem, in order to fix it, I had to comment out the <name .../> statement in the pom.xml file. Not sure whether this has anything to do with Quarkus, I reckon it doesn't, however I'm doing Maven since 10 years and I've never seen something like this.

I might register a new issue for that but, for now, the initial issue is closed and it was due to my typos which weren't flagged by the compiler as the erroneous artifacts existed, by coincidence.
